platforms/esp/32/fastpin_esp32.h

```diff
@@ -24,68 +11,90 @@ template<uint8_t PIN, uint32_t MASK> class _ESPPIN {
   inline static void setOutput() { pinMode(PIN, OUTPUT); }
   inline static void setInput() { pinMode(PIN, INPUT); }
 
-  inline static void hi() __attribute__ ((always_inline)) { if(PIN < OUTPUT_PIN_LIMIT) { _GPB0._GPOS = MASK; } }
-  // inline static void hi() __attribute__ ((always_inline)) { gpio_set_level((gpio_num_t)PIN, HIGH); }
+  inline static void hi() __attribute__ ((always_inline)) { 
+      if (PIN < 32) GPIO.out_w1ts = MASK;
+      else if(PIN < 34) GPIO.out1_w1ts.val = MASK;
+  }
+
+  inline static void lo() __attribute__ ((always_inline)) {
+      if (PIN < 32) GPIO.out_w1tc = MASK;
+      else if(PIN < 34) GPIO.out1_w1tc.val = MASK;
+  }
 
-  inline static void lo() __attribute__ ((always_inline)) { if (PIN < OUTPUT_PIN_LIMIT){ _GPB0._GPOC = MASK; } }
-  // inline static void lo() __attribute__ ((always_inline)) { gpio_set_level((gpio_num_t)PIN, LOW); }
-  inline static void set(register port_t val) __attribute__ ((always_inline)) { if (PIN < OUTPUT_PIN_LIMIT){ _GPB0._GPO = val; }}
-  // inline static void set(register port_t val) __attribute__ ((always_inline)) { gpio_set_level((gpio_num_t)PIN, val); }
+  inline static void set(register port_t val) __attribute__ ((always_inline)) {
+      if (val) hi();
+      else lo();
+  }
 
   inline static void strobe() __attribute__ ((always_inline)) { toggle(); toggle(); }
 
-  inline static void toggle() __attribute__ ((always_inline)) { if (PIN < OUTPUT_PIN_LIMIT){ _GPB0._GPO = MASK; } }
+  inline static void toggle() __attribute__ ((always_inline)) {  }
 
   inline static void hi(register port_ptr_t port) __attribute__ ((always_inline)) { hi(); }
   inline static void lo(register port_ptr_t port) __attribute__ ((always_inline)) { lo(); }
   inline static void fastset(register port_ptr_t port, register port_t val) __attribute__ ((always_inline)) { *port = val; }
 
-  inline static port_t hival() __attribute__ ((always_inline)) { if (PIN<OUTPUT_PIN_LIMIT) { return GPIO_OUT_REG | MASK;    }}
-  inline static port_t loval() __attribute__ ((always_inline)) { if (PIN<OUTPUT_PIN_LIMIT) { return GPIO_OUT_REG & ~MASK;   }}
-  inline static port_ptr_t port() __attribute__ ((always_inline)) { if(PIN<OUTPUT_PIN_LIMIT) { return &_GPB0._GPO;   }}
-  inline static port_ptr_t sport() __attribute__ ((always_inline)) { if (PIN<OUTPUT_PIN_LIMIT) {return &_GPB0._GPOS; }}
-  inline static port_ptr_t cport() __attribute__ ((always_inline)) { if (PIN<OUTPUT_PIN_LIMIT) {return &_GPB0._GPOC; }}
+  inline static port_t hival() __attribute__ ((always_inline)) { return MASK; }
```

Comment:
hival/loval are used as an optimization in bitbang'd spi output - to avoid a few rounds of bitwise operators.  hival is the "write this value to the I/o register to make this pin high" while loval is the "write this value to the I/o register to make the pin low" -- so the value that's returned is the full state of the I/o register, adjusted for whether we want this pin to be high or low - for example:

```
	inline static port_t hival() __attribute__ ((always_inline)) { return ((PIN<32) ? GPIO.out : GPIO.out1) | _MASK; }
	inline static port_t loval() __attribute__ ((always_inline)) { return ((PIN<32) ? GPIO.out : GPIO.out1)  & ~_MASK; }
```

Meanwhile port returns the register itself - as in a pointer to the register:

```
    inline static port_ptr_t port() ((always_inline)) { return (PIN < 32) ? &GPIO.out : &GPIO.out1); }
```

Reply:
The way it seems to work on the ESP32 is that you use the same mask regardless of whether you want to set a pin high or low -- the difference is *which* register you write that mask to. That is, hival and lowval are the same, the difference is sport vs cport. The port() method will never work for this reason.

Reply:
Yes - most higher end arm/32 bit cores have all _three_ types of register - a 32-bit register that gives the current hi/lo values for all 32 pins on that port, a 32-bit register that, when you write to it turns on any pin that has a corresponding 1 bit in the written value, and a 32-bit register that, when you write to it, turns _off_ any pin that has a corresponding 1 bit in the written value. - for the ESP32 there's GPIO.out - which is the port register - which, if I understand correctly you _can_ use to read the state of an output only pin (whether it is hi or low) - as well as bulk set the value (both hi and low) of multiple pins at the same time.  This is along side the out_w1tc which is the "write 1's to clear pins" and the out_w1ts "write 1's to set pins" (the c and s suffixes on those make it clear which one they are)- so, if you use GPIO.out then hival/lowval and port _will_ in fact work the way that the rest of the library is expecting it to.

From the gpio_struct header file:

```
    uint32_t out;                                   /*GPIO0~31 output value*/
    uint32_t out_w1ts;                              /*GPIO0~31 output value write 1 to set*/
    uint32_t out_w1tc;                              /*GPIO0~31 output value write 1 to clear*/
```

There are also a variety of other possible port options.  Some arm platforms have a toggle register, where any bit  in the 32-bit value that you write that has a 1 will flip the state of the associated pin (hi to low or low to high).  Then there's also yet another register writing technique called "bit-banding" - where a virtual address space is exploded out so that each pin has 4 32-bit memory locations - an out, a set, a clear, and a toggle - and writing any value will do the appropriate action with the pin - the nice thing about this is _any_ non-zero value triggers the action (setting the value, setting the pin hi, setting it low, toggling it) - which is a nice way to short circuit the whole "if (someValue != 0)  GPIO.set = MASK; else GPIO.clear = MASK;" - you can just do bitband[pin].val = someValue;

So - yes, port, hival, and lowval can work the way it does for other platforms.  The GPIO.out register is how, among other things, you would ask "Is pin 24 hi or low right now?" (the w1tc and w1ts registers are effectively write only, there's nothing meaningful to read from them).

Reply:
Reading up on this question: it is definitely possible to implement all the functions by accessing the GPIO.out register. The only downside I see (which is very minor) is that it is slower because it requires three instructions: read, bit-op, write. Using the GPIO.out_wtx registers requires only a write.

Reply:
Right - which is why hi should use the out_w1ts and lo should use the out_w1tc - but port/hival/lowval specifically work with the out register.  There's other code in FastLED (and other private code I have that uses FastPin) that makes use of this - and again, something low level like FastPin I want to work reliably the same way across all the platforms, which is why i'm being a stickler about this.